src/nanoemoji/paint.py

```diff
@@ -111,21 +143,96 @@ class PaintRadialGradient(Paint):
     c1: Point = Point()
     r0: float = 0.0
     r1: float = 0.0
-    affine2x2: Optional[Tuple[float, float, float, float]] = None
 
     def colors(self):
         for stop in self.stops:
             yield stop.color
 
     def to_ufo_paint(self, colors):
-        result = {
+        paint = {
             "format": self.format,
             "colorLine": _ufoColorLine(self, colors),
             "c0": self.c0,
             "c1": self.c1,
             "r0": self.r0,
             "r1": self.r1,
         }
-        if self.affine2x2:
-            result["transform"] = self.affine2x2
-        return result
+        return paint
+
+
+@dataclasses.dataclass(frozen=True)
+class PaintGlyph(Paint):
+    format: ClassVar[int] = 4
+    glyph: str
+    paint: Paint
+
+    def colors(self):
+        yield from self.paint.colors()
+
+    def to_ufo_paint(self, colors):
+        paint = {
+            "format": self.format,
+            "glyph": self.glyph,
+            "paint": self.paint.to_ufo_paint(colors),
+        }
+        return paint
+
+
+@dataclasses.dataclass(frozen=True)
+class PaintColrGlyph(Paint):
+    format: ClassVar[int] = 5
+    glyph: str
+    first_layer_index: int
+    last_layer_index: int
+
+    def to_ufo_paint(self, _):
+        paint = {
+            "format": self.format,
+            "glyph": self.glyph,
+            "firstLayerIndex": self.first_layer_index,
+            "lastLayerIndex": self.last_layer_index,
+        }
+        return paint
+
+
+@dataclasses.dataclass(frozen=True)
+class PaintTransform(Paint):
+    format: ClassVar[int] = 6
+    transform: Tuple[float, float, float, float, float, float]
+    paint: Paint
+
+    def colors(self):
+        yield from self.paint.colors()
+
+    def to_ufo_paint(self, colors):
+        paint = {
+            "format": self.format,
+            "transform": self.transform,
+            "paint": self.paint.to_ufo_paint(colors),
+        }
+        return paint
+
+
+@dataclasses.dataclass(frozen=True)
+class PaintComposite(Paint):
+    format: ClassVar[int] = 7
+    mode: CompositeMode
+    source: Paint
+    backdrop: Paint
+
+    def colors(self):
+        yield from self.source.colors()
+        yield from self.backdrop.colors()
+
+    def to_ufo_paint(self, colors):
+        paint = {
+            "format": self.format,
+            "mode": self.mode.name.lower(),
+            "source": self.source.to_ufo_paint(colors),
+            "backdrop": self.backdrop.to_ufo_paint(colors),
+        }
+        return paint
+
+@dataclasses.dataclass(frozen=True)
+class PaintColrLayers(Paint):
+    format: ClassVar[int] = 8
```

Comment:
no newline at EOF, my vim always forces me to add one

Reply:
black is removing it when I add one